### PR TITLE
Query Cache: Include settings into query cache key

### PR DESCRIPTION
**Pull request by rschu1ze**
@Algunenano [suggested](https://github.com/ClickHouse/ClickHouse/pull/64199#discussion_r1608536243) that query cache keys should include session/query-level settings. That makes sense, e.g. for settings [additional_table_filters](https://clickhouse.com/docs/en/operations/settings/settings#additional_table_filters) and [limit](https://clickhouse.com/docs/en/operations/settings/settings#limit) (\+ possibly more) which can affect the query result.

(This PR also fixes a minor glitch in https://github.com/ClickHouse/ClickHouse/pull/62696, "FunctionHelpers.cpp")

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The query cache now considers identical queries with different settings as different. This increases robustness in cases where different settings (e.g. `limit` or `additional_table_filters`) would affect the query result.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

</details>

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit ea93fa6b2ca35a7c86579bdc5c455454a0bcfd9b with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64205/ea93fa6b2ca35a7c86579bdc5c455454a0bcfd9b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Functions/FunctionHelpers.cpp [hunk 1]
```diff
@@ -21,8 +21,6 @@ namespace ErrorCodes
 
 const ColumnConst * checkAndGetColumnConstStringOrFixedString(const IColumn * column)
 {
-    if (!column)
```
Comment (pufit) [inline src/Functions/FunctionHelpers.cpp:24]: why do we remove this check?

**Review by pufit — APPROVED** [review]

File: src/Functions/FunctionHelpers.cpp (hunk 1 above)
Comment (pufit) [reply src/Functions/FunctionHelpers.cpp:24]: ok, nevermind, I see https://github.com/ClickHouse/ClickHouse/pull/62696

**Comment by rschu1ze** [comment]
02494_query_cache_nested_query_bug [failed](https://s3.amazonaws.com/clickhouse-test-reports/64205/6e6e2944b56245cd5eefd14deb7dba7b8459b935/stateless_tests__release_.html) and after some debugging, I uncovered some serious fuckup.

The test runs an expensive aggregation query twice and it expects that the result of the second invocation is delivered from the cache. Each query invocation does two things (given that the query is run with `SETTINGS use_query_cache = 1`):
1. It checks if the query result is cached. If yes --> read the result from the cache, return it to the user, and be done.
2. After executing the query, write the query result into the cache.

With this PR, settings become part of the query entry key. Funnily, steps 1 and 2 suddenly generate different hashes. The reason is that the settings in step 2 include `use_structure_from_insertion_table_in_table_functions = 0`. It is set during query execution between steps 1 and 2 in https://github.com/ClickHouse/ClickHouse/blob/de2a0be02580eb7501ad1fec5de35a7107ef9a1e/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp#L85. As a result, the first query invocation inserts a cache entry with hash X but the second query invocation uses a different hash Y, does not find the entry, and does the aggregation again.

While messing with settings internally seems questionable in the first place, I am afraid this could happen with other settings and in other places too. This would silently render there query cache useless 😢

So instead of a whitelist approach, I'lll go for a blacklist approach - explicitly check for settings that will (likely) alter the query result. I actually wanted to avoid this approach but it seems I have no choice.

**Pull request merged**